### PR TITLE
Add Support for Scala 2.13 in Spark 3.4.1

### PR DESCRIPTION
**Pull request by galacticgumshoe**
### What changes were proposed in this pull request? 

Add a Scala 2.13 Dockerfile to the Spark 3.4.1 sub-folder for building and tagging `3.4.1-scala2.13-java11-ubuntu`, `3.4.1-scala2.13`, `scala2.13` image.

### Why are the changes needed? 

To provide support for projects using Scala 2.13 and Spark 3.4.1.

### Does this PR introduce _any_ user-facing change? 

Yes, this would add new Docker tags: `3.4.1-scala2.13-java11-ubuntu`, `3.4.1-scala2.13`, `scala2.13`.

### How was this patch tested? 

1. Ran `./add-dockerfiles.sh 3.4.1` in my local
2. Once I verified the appropriate directory, Dockerfile, and entrypoint.sh were generated in `3.4.1/scala2.13-java11-ubuntu` I then built an image in my local: `docker build --platform linux/amd64 -t <image>:<tag> 3.4.1/scala2.13-java11-ubuntu/`
3. Verified the correct binary was downloaded in the 2.13 image: https://archive.apache.org/dist/spark/spark-3.4.1/spark-3.4.1-bin-hadoop3-scala2.13.tgz.
4. Ran spark-shell and verified versions (2.13): `docker run -it <image>:<tag> /opt/spark/bin/spark-shell`
> Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.4.1
      /_/
Using Scala version 2.13.8 (OpenJDK 64-Bit Server VM, Java 11.0.19)
5. I also built the scala2.12 image in my local: `docker build --platform linux/amd64 -t <image>:<tag> 3.4.1/scala2.12-java11-ubuntu/`.
7. Verified the correct binary was downloaded in the 2.12 image: https://archive.apache.org/dist/spark/spark-3.4.1/spark-3.4.1-bin-hadoop3.tgz
8. Ran spark-shell and verified versions (2.12): `docker run -it <image>:<tag> /opt/spark/bin/spark-shell`
>Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.4.1
      /_/
Using Scala version 2.12.17 (OpenJDK 64-Bit Server VM, Java 11.0.19)

### CODE REVIEW

File: add-dockerfiles.sh [hunk 1]
```diff
@@ -33,6 +33,7 @@ scala2.12-java11-python3-r-ubuntu
 scala2.12-java11-python3-ubuntu
 scala2.12-java11-r-ubuntu
 scala2.12-java11-ubuntu
+scala2.13-java11-ubuntu
```
Comment (Yikun) [inline add-dockerfiles.sh:36]: You only add the scala one, do you think python3/r/all should also be added in sometime? (Yep, just a question, IMO we can add scala2.13 now, and add others on demand in future)

**Review by Yikun — COMMENTED** [review]
Thanks for contributions! I left some comments.

You might also want to add the test for this image.
https://github.com/apache/spark-docker/blob/master/.github/workflows/build_3.4.1.yaml#L29

```
jobs:
  run-build:
    strategy:
      matrix:
        image-type: ["all", "python", "scala", "r"]
      include:
        - spark: 3.4.1
          scala: 2.12
          java: 11
          image-type: scala
    name: Run
    secrets: inherit
    uses: ./.github/workflows/main.yml
    with:
      spark: 3.4.1
      scala: 2.12
      java: 11
      image-type: ${{ matrix.image-type }}
```

File: Dockerfile.template [hunk 2]
```diff
@@ -36,15 +36,16 @@ RUN set -ex; \
 
 # Install Apache Spark
 # https://downloads.apache.org/spark/KEYS
-ENV SPARK_TGZ_URL=https://archive.apache.org/dist/spark/spark-{{ SPARK_VERSION }}/spark-{{ SPARK_VERSION }}-bin-hadoop3.tgz \
-    SPARK_TGZ_ASC_URL=https://archive.apache.org/dist/spark/spark-{{ SPARK_VERSION }}/spark-{{ SPARK_VERSION }}-bin-hadoop3.tgz.asc \
```
[the remark this answers is not in the document]
Comment (galacticgumshoe) [reply Dockerfile.template:40]: Makes sense, will update.

File: 3.4.1/scala2.12-java11-ubuntu/Dockerfile [hunk 3]
```diff
@@ -36,15 +36,16 @@ RUN set -ex; \
...
 # Install Apache Spark
 # https://downloads.apache.org/spark/KEYS
-ENV SPARK_TGZ_URL=https://archive.apache.org/dist/spark/spark-3.4.1/spark-3.4.1-bin-hadoop3.tgz \
-    SPARK_TGZ_ASC_URL=https://archive.apache.org/dist/spark/spark-3.4.1/spark-3.4.1-bin-hadoop3.tgz.asc \
+ENV SPARK_TGZ_URL=https://archive.apache.org/dist/spark/spark-3.4.1/spark-3.4.1-bin- \
+    SPARK_TGZ_ASC_URL=https://archive.apache.org/dist/spark/spark-3.4.1/spark-3.4.1-bin- \
     GPG_KEY=F28C9C925C188C35E345614DEDA00CE834F0FC5C
 
 RUN set -ex; \
+    if [ "2.12" = "2.13" ]; then export BIN_FILE_SUFFIX="hadoop3-scala2.13.tgz"; else export BIN_FILE_SUFFIX="hadoop3.tgz"; fi; \
```
[the remark this answers is not in the document]
Comment (galacticgumshoe) [reply 3.4.1/scala2.12-java11-ubuntu/Dockerfile:44]: Agreed.

File: versions.json [hunk 4]
```diff
@@ -22,8 +22,16 @@
       "path": "3.4.1/scala2.12-java11-ubuntu",
       "tags": [
         "3.4.1-scala2.12-java11-ubuntu",
-        "3.4.1-scala",
-        "scala"
+        "3.4.1-scala2.12",
+        "scala2.12"
```
[the remark this answers is not in the document]
Comment (galacticgumshoe) [reply versions.json:26]: Will fix!

File: add-dockerfiles.sh (hunk 1 above)
Comment (galacticgumshoe) [reply add-dockerfiles.sh:36]: I didn't see any other bin file that was compiled with Scala 2.13 in the [Apache Spark Archives](https://archive.apache.org/dist/spark/spark-3.4.1/). So I thought to hold off on trying the others. But now that I see they are built on top of the initial image (not from the Dockerfile.template, but from the r-python.template), I can try to add support in this PR.

**Comment by galacticgumshoe** [comment]
@Yikun There, I think that should address each of your comments. Let me know otherwise. Thanks for taking a look at this!

**Comment by galacticgumshoe** [comment]
@Yikun Back from vacation, so just catching up again here. It looks like the failure in the workflow stems from this file not found: 

`Exception in thread "main" java.nio.file.NoSuchFileException: /opt/spark/examples/jars/spark-examples_2.12-3.4.1.jar`

I am trying to follow the workflow logic, but some of these dependencies relating to Kubernetes look like they are not controlled by this project. Can you provide any insights on this?

**Comment by Yikun** [comment]
The CI should be fixed by https://github.com/apache/spark-docker/pull/53 , you can rebase after it merge.

File: 3.4.1/scala2.12-java11-ubuntu/Dockerfile [hunk 5]
```diff
@@ -43,8 +43,8 @@ ENV SPARK_TGZ_URL=https://archive.apache.org/dist/spark/spark-3.4.1/spark-3.4.1-
 RUN set -ex; \
     export SPARK_TMP="$(mktemp -d)"; \
     cd $SPARK_TMP; \
-    wget -nv -O spark.tgz "$SPARK_TGZ_URL"; \
-    wget -nv -O spark.tgz.asc "$SPARK_TGZ_ASC_URL"; \
+    wget -nv -O spark.tgz "${SPARK_TGZ_URL}"; \
+    wget -nv -O spark.tgz.asc "${SPARK_TGZ_ASC_URL}"; \
```
Comment (Yikun) [inline 3.4.1/scala2.12-java11-ubuntu/Dockerfile:47]: Is it a required change? otherwise I think we might better to recover.

File: .github/workflows/build_3.4.1_2.13.yaml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  run-build:
+    strategy:
+      matrix:
+        image-type: ["all", "python", "scala", "r"]
+    name: Run
+    secrets: inherit
+    uses: ./.github/workflows/main.yml
+    with:
+      spark: 3.4.1
+      scala: 2.13
```
Comment (Yikun) [inline .github/workflows/build_3.4.1_2.13.yaml:39]: I think we can move `scala: 2.13` to strategy and squash this file to build_3.4.1.yaml:

```
    strategy:
      matrix:
        image-type: ["all", "python", "scala", "r"]
        scala: ["2.13", "2.12"]
```

Comment (galacticgumshoe) [reply .github/workflows/build_3.4.1_2.13.yaml]: Should be fixed now.

File: 3.4.1/scala2.12-java11-ubuntu/Dockerfile (hunk 5 above)
Comment (galacticgumshoe) [reply 3.4.1/scala2.12-java11-ubuntu/Dockerfile:47]: So while not technically required, this is a result of the changes in the PR to the Dockerfile.template following running the add-docker files.sh script against 3.4.1. I will back out the brackets in the template and rerun to remove these brackets from the generated Dockerfile here. Just documenting the cause of this change.

Comment (galacticgumshoe) [reply 3.4.1/scala2.12-java11-ubuntu/Dockerfile:47]: Please note that since I merged in the changes from master in 81c2933 that had merged after this PR was initiated, that pulled in changes also to the entrypoint.sh.template. So after making the change to the Dockerfile.template and rerunning add-dockerfiles.sh, both the Dockerfile and entrypoint.sh for 3.4.1/2.12 has been changed.

**Comment by galacticgumshoe** [comment]
@Yikun After syncing my fork with the changes merged to master branch since this PR initiated, I believe this is ready for a review. However, considering that v3.3.3 and 3.5.0 were added recently, would it make sense to include support of Scala 2.13 to these latest versions as well? Or all versions?

File: 3.4.1/scala2.12-java11-ubuntu/entrypoint.sh [hunk 7]
```diff
@@ -77,6 +77,9 @@ elif ! [ -z "${SPARK_HOME+x}" ]; then
   SPARK_CLASSPATH="$SPARK_HOME/conf:$SPARK_CLASSPATH";
 fi
 
+# SPARK-43540: add current working directory into executor classpath
```
Comment (Yikun) [inline 3.4.1/scala2.12-java11-ubuntu/entrypoint.sh:80]: This is 3.5.x series, we might create a `entrypoint.sh.3.4.template`, `Dockerfile.3.4.template`.

We also [change](https://github.com/apache/spark-docker/blob/master/add-dockerfiles.sh#L53) the `add-dockerfiles.sh`, For 3.x.1 version, if `3.x` templates exists, use 3.x template, otherwise use the master entrypoint and Dockerfile directly.

It can be a separate PR.

Comment (Yikun) [reply 3.4.1/scala2.12-java11-ubuntu/entrypoint.sh:80]: cc @HyukjinKwon @zhengruifeng @dongjoon-hyun 

Or we could just apply master dockerfile changes in 3.4-branches?

File: 3.4.1/scala2.12-java11-ubuntu/entrypoint.sh [hunk 8]
```diff
@@ -90,6 +93,7 @@ case "$1" in
     CMD=(
       "$SPARK_HOME/bin/spark-submit"
       --conf "spark.driver.bindAddress=$SPARK_DRIVER_BIND_ADDRESS"
+      --conf "spark.executorEnv.SPARK_DRIVER_POD_IP=$SPARK_DRIVER_BIND_ADDRESS"
```
Comment (Yikun) [inline 3.4.1/scala2.12-java11-ubuntu/entrypoint.sh:96]: ditto

File: versions.json [hunk 9]
```diff
@@ -40,18 +40,41 @@
...
+    },
     {
       "path": "3.4.1/scala2.12-java11-ubuntu",
       "tags": [
         "3.4.1-scala2.12-java11-ubuntu",
-        "3.4.1-scala"
+        "3.4.1-scala2.12",
+        "scala2.12",
+        "3.4.1-scala",
+        "scala"
```
Comment (Yikun) [inline versions.json:69]: scala should be removed, because it's 3.5.0 now

File: versions.json [hunk 10]
```diff
@@ -40,18 +40,41 @@
...
+        "3.4.1-scala2.13-java11-r-ubuntu"
+      ]
+    },
     {
       "path": "3.4.1/scala2.12-java11-ubuntu",
       "tags": [
         "3.4.1-scala2.12-java11-ubuntu",
-        "3.4.1-scala"
+        "3.4.1-scala2.12",
+        "scala2.12",
```
Comment (Yikun) [inline versions.json:67]: scala2.12 should be removed, because it's 3.5.0 now

**Review by Yikun — COMMENTED** [review]
Sorry for the late reply, I just came back from holidays

> However, considering that v3.3.3 and 3.5.0 were added recently, would it make sense to include support of Scala 2.13 to these latest versions as well? Or all versions?

I prefer to only add the new version support in latest version, that is 3.5.0.

File: versions.json [hunk 11]
```diff
@@ -40,18 +40,41 @@
...
+        "3.4.1-scala",
+        "scala"
+      ]
+    },
+    {
+      "path": "3.4.1/scala2.13-java11-ubuntu",
+      "tags": [
+        "3.4.1-scala2.13-java11-ubuntu",
+        "3.4.1-scala2.13",
+        "scala2.13"
```
Comment (holdenk) [inline versions.json:77]: Would we want to do the scala 2.13 tag with 3.5 instead  @Yikun

Comment (Yikun) [reply versions.json:77]: Yes, agree! As I mentioned above comment, 2.13 better to introduce in latest versio (3.5.0 for now)

https://github.com/apache/spark-docker/pull/52#pullrequestreview-1676547189

File: 3.4.1/scala2.12-java11-ubuntu/entrypoint.sh (hunk 7 above)
Comment (HyukjinKwon) [reply 3.4.1/scala2.12-java11-ubuntu/entrypoint.sh:80]: I think creating a template is better.

**Comment by Philosh** [comment]
how do I install pyspark 3.4.1 with scala 2.13? please help. I only get scala 2.12

**Comment by caldempsey** [comment]
Also running into this issue

**Comment by databius** [comment]
It would be great if we could support old versions instead of only spark 3.5+.
I need an image that supports scala 2.13 and spark 3.4.2.
Currently, I am building my own image based on this PR. Hopefully the official image will be published soon.
